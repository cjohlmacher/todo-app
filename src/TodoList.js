import React, { useState } from 'react';
import {v4 as uuid } from 'uuid';
import './TodoList.css';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';
import List from '@mui/material/List';


const TodoList = () => {
    const [todos, setTodos] = useState([]);
    
    const todoComponents = todos.map(({id,description}) => {
        return <Todo key={id} description={description} deleteTodo={() => deleteTodo(id)}/>
    });

    const deleteTodo = (id) => {
        setTodos((todos) => todos.filter((todo) => todo.id !== id));
    };

    const addTodo = ({description}) => {
        const id = uuid();
        setTodos((todos) => [...todos, {id, description}]);
    };

    return (
        <>
            <List>
                {todoComponents}
            </List>
            <NewTodoForm addTodo={addTodo}/>
        </>
    )
};

export default TodoList;