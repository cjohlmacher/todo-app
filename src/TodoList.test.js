import { fireEvent, render, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import TodoList from './TodoList';

const addTodo = async (getByLabelText, getByRole) => {
    const todoInput = getByLabelText('Add Todo:');
    const addButton = getByRole('button');
    fireEvent.change(todoInput, {target: {name: 'todo', value: 'Wash dishes'}});
    fireEvent.click(addButton);
};

it('should render without breaking', () => {
    render(<TodoList />);
});

it('should match the snapshot', () => {
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it('should add a todo on submit', async () => {
    const { getByLabelText, getByRole, getByText } = render(<TodoList />);
    addTodo(getByLabelText, getByRole);
    await waitFor( () => { 
        const newTodo = getByText("Wash dishes");
        expect(newTodo).toBeInTheDocument();
    });
});

it('should match the snapshot after adding todo', async () => {
    const { getByLabelText, getByRole, getByText, asFragment } = render(<TodoList />);
    addTodo(getByLabelText, getByRole);
    await waitFor( () => {
        const newTodo = getByText("Wash dishes");
        expect(newTodo).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
});

it('should delete a todo on delete click', async () => {
    const { getByLabelText, getByRole, getByText, queryByText, getByTestId } = render(<TodoList />);
    addTodo(getByLabelText, getByRole);
    await waitFor( () => {
        const newTodo = getByText("Wash dishes");
        expect(newTodo).toBeInTheDocument();
    });
    const deleteButton = getByTestId('DeleteOutlinedIcon');
    fireEvent.click(deleteButton);
    await waitFor( () => {
        const newTodo = queryByText("Wash dishes");
        expect(newTodo).toBe(null);
    });
    const newTodo = queryByText("Wash dishes");
    expect(newTodo).not.toBeInTheDocument();
});