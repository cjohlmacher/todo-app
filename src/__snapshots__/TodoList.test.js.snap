// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  />
  <div>
    <div
      class="MuiBox-root css-gmuwbf"
    >
      <form>
        <label
          for="todo"
        >
          Add Todo:
        </label>
        <div
          class="MuiFormControl-root MuiTextField-root css-ewmcuh-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="todo"
              name="todo"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-nnbavb"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          aria-label="button"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-92r701-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            name="add"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match the snapshot after adding todo 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div
      class="MuiBox-root css-361be1"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      >
        <div
          class="MuiListItemText-root css-igjnpd-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Wash dishes
          </span>
        </div>
        <button
          aria-label="button"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1d159sf-MuiSvgIcon-root"
            data-testid="DeleteOutlinedIcon"
            focusable="false"
            name="delete"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 9v10H8V9h8m-1.5-6h-5l-1 1H5v2h14V4h-3.5l-1-1zM18 7H6v12c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
    </div>
  </ul>
  <div>
    <div
      class="MuiBox-root css-gmuwbf"
    >
      <form>
        <label
          for="todo"
        >
          Add Todo:
        </label>
        <div
          class="MuiFormControl-root MuiTextField-root css-ewmcuh-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="todo"
              name="todo"
              type="text"
              value="Wash dishes"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-nnbavb"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          aria-label="button"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-92r701-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            name="add"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
